pkg/virt-controller/watch/vm.go

```diff
@@ -917,24 +933,34 @@ func (c *VMController) startStop(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualM
 		// For this RunStrategy, a VMI should only be restarted if it failed.
 		// If a VMI enters the Succeeded phase, it should not be restarted.
 		if vmi != nil {
-			var forceStop bool
-			if forceStop = hasStopRequestForVMI(vm, vmi); forceStop {
+			forceStop := hasStopRequestForVMI(vm, vmi)
+			if forceStop {
 				log.Log.Object(vm).Infof("processing stop request for VMI with phase %s and VM runStrategy: %s", vmi.Status.Phase, runStrategy)
-
 			}
+			vmiFailed := vmi.Status.Phase == virtv1.Failed
+			vmiSucceeded := vmi.Status.Phase == virtv1.Succeeded
 
-			if forceStop || vmi.Status.Phase == virtv1.Failed {
-				// For RerunOnFailure, this controller should only restart the VirtualMachineInstance
-				// if it failed.
+			if vmi.DeletionTimestamp == nil && (forceStop || vmiFailed || vmiSucceeded) {
```

Comment:
This is a behavior change, we weren't deleting succeeded VMIs before, is there a reason to prefer this over keeping them?

It can be confusing not to see the VMI anymore if it terminated without errors, can't we keep it around until there's a new start request?

Reply:
Keeping successful VMIs around is actually a hack. The presence of the finished VMI object is used as a boolean status to tell the code not to start the VM again.
The ability to remove that is actually a nice side effect of the changes introduced by this PR!
Note: we already don't keep VMI objects around when using other run strategies.

Reply:
Ok thanks!